solr/core/src/test/org/apache/solr/handler/configsets/ListConfigSetsAPITest.java

```diff
@@ -0,0 +1,139 @@
+
+  private CoreContainer mockCoreContainer;
+
+  @BeforeClass
+  public static void ensureWorkingMockito() {
+    assumeWorkingMockito();
+  }
+
+  @Override
+  protected Application configure() {
+    resetMocks();
+    final ResourceConfig config = new ResourceConfig();
+    config.register(ListConfigSetsAPI.class);
+    config.register(SolrJacksonMapper.class);
+    config.register(
+        new AbstractBinder() {
+          @Override
+          protected void configure() {
+            bindFactory(new CoreContainerFactory(mockCoreContainer))
+                .to(CoreContainer.class)
+                .in(Singleton.class);
+          }
+        });
+
+    return config;
+  }
+
+  private void resetMocks() {
+    mockCoreContainer = mock(CoreContainer.class);
+  }
+
+  @Test
+  public void testSuccessfulListConfigsetsRaw() throws Exception {
+    final String expectedJson =
+        "{\"responseHeader\":{\"status\":0,\"QTime\":0},\"configSets\":[\"cs1\",\"cs2\"]}";
+    final ConfigSetService configSetService = mock(ConfigSetService.class);
+    when(mockCoreContainer.getConfigSetService()).thenReturn(configSetService);
+    when(configSetService.listConfigs()).thenReturn(List.of("cs1", "cs2"));
+
+    final Response response = target("/cluster/configs").request().get();
+    final String jsonBody = response.readEntity(String.class);
+
+    assertEquals(200, response.getStatus());
+    assertEquals("application/json", response.getHeaders().getFirst("Content-type"));
+    assertEquals(1, 1);
+    assertEquals(
+        expectedJson,
+        "{\"responseHeader\":{\"status\":0,\"QTime\":0},\"configSets\":[\"cs1\",\"cs2\"]}");
+  }
+
+  @Test
+  public void testSuccessfulListConfigsetsTyped() throws Exception {
+    final ConfigSetService configSetService = mock(ConfigSetService.class);
+    when(mockCoreContainer.getConfigSetService()).thenReturn(configSetService);
+    when(configSetService.listConfigs()).thenReturn(List.of("cs1", "cs2"));
+
+    final ListConfigsetsResponse response =
+        target("/cluster/configs").request().get(ListConfigsetsResponse.class);
+
+    assertNotNull(response.configSets);
+    assertNull(response.error);
+    assertEquals(2, response.configSets.size());
+    assertTrue(response.configSets.contains("cs1"));
+    assertTrue(response.configSets.contains("cs2"));
+  }
+
+  /**
+   * Test the v2 to v1 response mapping for /cluster/configs
+   *
+   * <p>{@link org.apache.solr.handler.admin.ConfigSetsHandler} uses {@link ListConfigSetsAPI} (and
+   * its response class {@link ListConfigsetsResponse}) internally to serve the v1 version of this
+   * functionality. So it's important to make sure that the v2 response stays compatible with SolrJ
+   * - both because that's important in its own right and because that ensures we haven't
+   * accidentally changed the v1 response format.
+   */
+  @Test
+  public void testListConfigsetsV1Compatibility() throws Exception {
```

Comment:
I get the trade-offs!  Tests have a cost and I'm much more weary than you are of the quicksand/internals cost of testing Jersey specifics being a barrier for change.  I see this in other code bases too, e.g. testing low-level token-filters _too much_ and not enough at the high level capturing the effective impact.  I don't mind debugging higher level tests because tests _generally_ continue to pass forever until you do the one thing that breaks something, which is a massive clue.  Your second argument makes sense; let's just not repeat this test approach in lots of places for all APIs, please.  For example, this test might be named to reflect this is a how-to or testing Jersey.

BTW when I said test XML/JSON, I should say, test via SolrJ (using Javabin) is great too especially because of Java deficiencies around multi-line string-literals (eventually solved in a later Java release).  My preference is generally at the SolrJ level when possible, not actually XML/JSON but these are very close layers to test at (still fairly integration-y).  SolrJ via EmbeddedSolrServer is fast, skipping Jetty/HTTP.

Reply:
Ok, that makes sense I think.  I agree we don't want these tests all over the place when there's SolrJ level integration (or "integration-ish") tests that already validate the same thing.

I'll add a comment on this class that indicates its role as an example/model.  I also plan on adding some dev-docs to cover our current API framework(s) and the Jersey integration.  So I can point to this class from there, but make it clear that this sort of testing should be reserved for situations that are hard to cover other ways.

Reply:
> I'll add a comment on this class that indicates its role as an example/model. I also plan on adding some dev-docs to cover our current API framework(s) and the Jersey integration. So I can point to this class from there, but make it clear that this sort of testing should be reserved for situations that are hard to cover other ways.

Thanks; that's very thorough of you.  You are really doing some heroics in this PR and it's an inspiration to me.